vim-patch:8.0.{10{39,53,55},1274}: cannot change a line in not current buffer

blueyed opened the pull request:
Problem:    Cannot change a line in a buffer other than the current one.
Solution:   Add setbufline(). (Yasuhiro Matsumoto, Ozaki Kiichi, closes vim/vim#1953)
https://github.com/vim/vim/commit/b31cf2bb0be95d106bd8eef93cc07550591c1d0d

blueyed committed:
vim-patch:8.0.1039: cannot change a line in not current buffer

Problem:    Cannot change a line in a buffer other than the current one.
Solution:   Add setbufline(). (Yasuhiro Matsumoto, Ozaki Kiichi, closes vim/vim#1953)
https://github.com/vim/vim/commit/b31cf2bb0be95d106bd8eef93cc07550591c1d0d

KillTheMule commented:
Shouldn't we implement this in terms of `nvim_buf_set_lines`? Seems wasteful to have 2 different implementations for the same thing, although it would ofc make porting future patches easier.

blueyed commented:
Yeah, I've thought that it is better to stick with Vim/upstream in this regard.
But have not really checked how easy it would be to use `nvim_buf_set_lines` for it yet.

blueyed commented on test/functional/eval/null_spec.lua:
```diff
@@ -61,7 +61,7 @@ describe('NULL', function()
       eq({''}, curbufmeths.get_lines(0, -1, false))
     end)
-    null_expr_test('does not crash setline()', 'setline(1, L)', 0, 0, function()
+    null_expr_test('does not crash setline()', 'setline(1, L)', 0, 1, function()
```
It changes the behavior of `:echo setline(1, v:_null_list)` - which is fine I guess, and rather constructed, isn't it?

bfredl replied:
changes it from what to what? Ideally, a null list should behave like a (locked) empty list.

blueyed replied:
From 0 to 1 - i.e. it returns "error" instead of "ok".

> Ideally, a null list should behave like a (locked) empty list.

Ok, but that's behavior different from Vim then.
Due to the check in https://github.com/neovim/neovim/pull/10250/files#diff-deb695d507653839019814f54242d47bR14878.

blueyed replied:
Previously it was using `if (argvars[1].v_type == VAR_LIST) {`.

Happy to keep it (and forward it to Vim), but it's the first I hear about null-lists.. :)

bfredl replied:
> Ok, but that's behavior different from Vim then.

We shouldn't regress from improvements made in nvim. Vim displays "null" list as if they were empty lists, which is their intentional behavior. The only reason one needs to "hear" about null list is because it behaves differently in some edge cases for no principal reason.

blueyed replied:
> Vim displays "null" list as if they were empty lists

But here it will cause a failure, i.e. it is not handled as if it is an empty list at least.
So it could be considered a bug there then.

As for the behavior in nvim it looks more like a side effect of not crashing on them.
check ceb45a088 / f572bd7e4e. (https://github.com/neovim/neovim/pull/10250/commits/b7584360705ef53a1555a2b7b5410c8339f9e08c)

bfredl replied:
> As for the behavior in nvim it looks more like a side effect of not crashing on them.

What does this even mean? In what way is accepting the list more "a side effect of not crashing" than throwing an error would be? How would this motivate one (non-crashing) behavior over another?

blueyed replied:
From @ZyX-I's commits it looked like he first refactored it, and when it crashed fixed that - but not using the condition that Vim had originally.

bfredl replied:
Sure, but the general strategy he followed was to make null list behave consistently like locked empty lists, rather than just pick any non-crashing behavior arbitrarily.

blueyed replied:
Ok, reverted.

blueyed commented:
Squashed, old head: 7f893eb.

blueyed committed:
vim-patch:8.0.1053: setline() does not work on startup

Problem:    setline() does not work on startup. (Manuel Ortega)
Solution:   Do not check for ml_mfp to be set for the current buffer.
            (Christian Brabandt)
https://github.com/vim/vim/commit/9d954207e2cc807b475bb04f8b59ef5bb3772d99

blueyed commented:
Added vim-patch:8.0.1053.

blueyed committed:
vim-patch:8.0.1055: bufline test hangs on MS-Windows

Problem:    Bufline test hangs on MS-Windows.
Solution:   Avoid message for writing file.  Source shared.vim when running
            test individually.
https://github.com/vim/vim/commit/11aa62f8f949bb590b4d7792a334885fba5e4137

blueyed committed:
vim-patch:8.0.1274: setbufline() fails when using folding

Problem:    setbufline() fails when using folding.
Solution:   Set "curwin" if needed. (Ozaki Kiichi, closes vim/vim#2293)
https://github.com/vim/vim/commit/0c4dc88a637a5027209aa00226996af84e248636

janlazo commented:
Don't forget https://github.com/neovim/neovim/pull/10059#issuecomment-496024694.

blueyed commented:
> Don't forget [#10059 (comment)](https://github.com/neovim/neovim/pull/10059#issuecomment-496024694).

What do you mean?  That 8.1.0039 was only partially merged in 237cecd81?
Would rather do that separate then, since this is rather big already.

blueyed commented:
@janlazo 
Please reply to my previous comment.

janlazo commented:
> > Don't forget [#10059 (comment)](https://github.com/neovim/neovim/pull/10059#issuecomment-496024694).
> 
> What do you mean? That 8.1.0039 was only partially merged in [237cecd](https://github.com/neovim/neovim/commit/237cecd81b83dcbdac05b208ea6adc3357039858)?
> Would rather do that separate then, since this is rather big already.

It is partially merged because some 8.1.x patch needed it and a PR with that 8.1.x patch can't be postponed. I prefer related patches to be done in one PR to handle all the edge cases, especially memory leaks and crashes (like 8.0.0286, 8.0.0493), because the code and tests are related.
